### PR TITLE
[android][camera] Add cornerPoints to barcode scanning result

### LINKED ISSUE #17795: [expo-camera] Don't receive cornerPoints / bounds on Android
### Summary

In the documentation for expo-camera for Expo SDK 45 it says that the [onBarCodeScanned](https://docs.expo.dev/versions/v45.0.0/sdk/camera/#onbarcodescanned) callback now returns a BarCodeScanningResult. This should (in most cases as I understand) also contain the cornerPoints (and bounds). In the linked snack below I won't get those cornerPoints using a Samsung A51.
Tried it on an iPhone and that returns the cornerPoints and bounds as expected.

I also checked the BarCodeScanner module with my Android phone and I do get cornerPoints and bounds.

I've also had a look at the source code, but I'm not that familiar with Kotlin to find the problem.

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

managed

### What platform(s) does this occur on?

Android

### SDK Version (managed workflow only)

45

### Environment

  expo-env-info 1.0.3 environment info:
    System:
      OS: Linux 5.10 Manjaro Linux
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 18.2.0 - /usr/bin/node
      Yarn: 1.22.19 - /usr/bin/yarn
      npm: 8.5.5 - /usr/bin/npm
    SDKs:
      Android SDK:
        API Levels: 29, 30, 31
        Build Tools: 29.0.2, 29.0.3, 30.0.0
        System Images: android-30 | Google APIs Intel x86 Atom, android-30 | Google Play Intel x86 Atom
    IDEs:
      Android Studio: AI-212.5712.43.2112.8609683
    npmPackages:
      expo: ~45.0.0 => 45.0.5 
      react: 17.0.2 => 17.0.2 
      react-dom: 17.0.2 => 17.0.2 
      react-native: 0.68.2 => 0.68.2 
      react-native-web: 0.17.7 => 0.17.7 
    npmGlobalPackages:
      expo-cli: 5.4.8
    Expo Workflow: managed

### Reproducible demo

https://snack.expo.dev/@gerrit_w/no-cornerpoints-on-android

### PR DESCRIPTION
**Pull request by igoro00**
# Why

Will fix #17795.

# How

`cornerPoints` wasn't implemented in the barcode scanner in `expo-camera` on Android. I implemented it with `expo-barcode-scanner` as a reference.

# Test Plan

I don't think there is a way to write an automated unit test for this kind of feature.

I modified the `bare-expo` app to show a bounding box instead of an Alert, like in BarCodeScanner screen. You can test it by turning on the barcode scanner in the Camera screen. Then point at some QR codes and check if the green bounding box matches the QR code on the camera. Repeat the test for both front and back camera and all orientations.

I tested it on Samsung Galaxy S20 5G (Android 12) and it works. I don't have any other working Android devices right now so I would appreciate more testing on different devices. 

# Checklist

- [x] Make it API-compatible - group number[] into {x: number, y: number}[]
- [x] Scale the values to view coordinates - [Reference](https://github.com/expo/expo/blob/main/android/versioned-abis/expoview-abi46_0_0/src/main/java/abi46_0_0/expo/modules/barcodescanner/BarCodeScannerView.kt#L71)
- [x] Make `cornerPoints` nullable and don't crash when `cornerPoints` are really not available.
- [x] Lint and format code with Expo rules
- [x] Update `bare-expo` 
- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).
^ I don't know. I can't get `expo prebuild` to complete without an error even on the main branch and this branch is no different.
- [x] Update Changelog

### CODE REVIEW

File: packages/expo-camera/android/src/main/java/expo/modules/camera/ExpoCameraView.kt [hunk 1]
```diff
@@ -189,13 +196,72 @@ class ExpoCameraView(
     barCodeScanner?.setSettings(settings)
   }
 
+  // Even = portrait, odd = landscape
+  private fun getDeviceOrientation() =
+          (context.getSystemService(Context.WINDOW_SERVICE) as WindowManager).defaultDisplay.rotation
```
Comment (lukmccall) [inline packages/expo-camera/android/src/main/java/expo/modules/camera/ExpoCameraView.kt:201]:
```suggestion
    (context.getSystemService(Context.WINDOW_SERVICE) as WindowManager).defaultDisplay.rotation
```

File: packages/expo-camera/android/src/main/java/expo/modules/camera/utils/BarCodeScannerListMappers.kt [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package expo.modules.camera.utils
+
+inline fun MutableList<Int>.mapY(block: (it: Int) -> Int) {
+    for (it in 0 until this.size step 2) {
+        this[it] = block(it)
+    }
```
Comment (lukmccall) [inline packages/expo-camera/android/src/main/java/expo/modules/camera/utils/BarCodeScannerListMappers.kt:6]:
```suggestion
  for (it in 0 until this.size step 2) {
    this[it] = block(it)
  }
```

File: packages/expo-camera/android/src/main/java/expo/modules/camera/utils/BarCodeScannerListMappers.kt [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+}
+
+inline fun MutableList<Int>.mapX(block: (it: Int) -> Int) {
+    for (it in 1 until this.size step 2) {
+        this[it] = block(it)
+    }
```
Comment (lukmccall) [inline packages/expo-camera/android/src/main/java/expo/modules/camera/utils/BarCodeScannerListMappers.kt:12]:
```suggestion
  for (it in 1 until this.size step 2) {
    this[it] = block(it)
  }
```

**Review by lukmccall — COMMENTED** [review]
Looks promising ;)

**Review by lukmccall — APPROVED** [review]
Thanks for your contribution.

I'll merge that. However, I think we should change the type of the `cornerPoints` property to not be optional. We should always return an empty array if we didn't get anything from the native code. What do you think about that? If you agree with me, I'll change that after the merge.

**Comment by igoro00** [comment]
I belive `Svg.Polygon`(imo the only use case for `cornerPoints`) doesn't care if you pass an empty array or undefined as `points`. So that would remove 1 or 2 `?` from code. If you want to do something different with that data, you can always check the length.

I honestly don't care. I didn't do it because I didn't want to change the API in my first contribution to the project. That would mean changing docs and more importantly making sure the iOS code never returns undefined as well(something I'm not willing to do. Going into the Android code was hard enough plus I don't own any Apple devices).

If you do it, I have nothing against that. On the Android side, you only need to revert [this commit](https://github.com/igoro00/expo/commit/4eae3898312c6706c21e1877bb4f29c6978626cc#diff-7d120ab165ddea0f148bf1fb94d2e9d4c950d52bd9730aca087d92de13041dca).

**Pull request merged**